File: C++/프로그래머스/2/42584. 주식가격/주식가격.cpp
#include <stack>
#include <vector>

using namespace std;

struct Price {
    int idx;
    int price;
};

vector<int> solution(vector<int> prices) {
    int N = prices.size();
    vector<int> answer(N);
    stack<Price> st;
    
    for(int i = 0; i < prices.size(); i++) {
        answer[i] = N - i - 1;
        while(!st.empty() &&prices[i] < st.top().price) {
            answer[st.top().idx] = i - st.top().idx;
            st.pop();
        }
        
        Price price;
        price.idx = i;
        price.price = prices[i];
        st.push(price);
    }
    
    
    return answer;
}

File: Python3/프로그래머스/2/42584. 주식가격/주식가격.py
def solution(prices):
    N = len(prices)
    answer = [i for i in range(N - 1, -1, -1)]
    stack = []
    
    time = 0
    for i, p in enumerate(prices):
        while stack and p < stack[-1][1]:
            # print("del", stack[-1], "in time", i - stack[-1][0])
            answer[stack[-1][0]] = i - stack[-1][0]
            del stack[-1]
        
        stack.append([i, p])
    
    return answer